keras_cv/ops/target_gather.py

```diff
@@ -0,0 +1,123 @@
+     3) target masking.
+
+    Args:
+     targets: [N, ...] or [batch_size, N, ...] Tensor representing targets such
+       as boxes, keypoints, etc.
+     indices: [M] or [batch_size, M] int32 Tensor representing indices within
+       `targets` to gather.
+     mask: optional [M, ...] or [batch_size, M, ...] boolean
+       Tensor representing the masking for each target. `True` means the corresponding
+       entity should be masked to `mask_val`, `False` means the corresponding entity
+       should be the target value.
+     mask_val: optinal float representing the masking value if `mask` is True on
+       the entity.
+
+     Returns:
+       targets: [M, ...] or [batch_size, M, ...] Tensor representing selected targets.
+
+     Raise:
+       ValueError: If `targets` is higher than rank 3.
+    """
+    targets_shape = targets.get_shape().as_list()
+    if len(targets_shape) > 3:
+        raise ValueError(
+            "`target_gather` does not support `targets` with rank "
+            "larger than 3, got {}".format(len(targets.shape))
+        )
+
+    def _gather_unbatched(labels, match_indices, mask, mask_val):
+        """Gather based on unbatched labels and boxes."""
+        num_gt_boxes = tf.shape(labels)[0]
+
+        def _assign_when_rows_empty():
+            if len(labels.shape) > 1:
+                mask_shape = [match_indices.shape[0], labels.shape[-1]]
+            else:
+                mask_shape = [match_indices.shape[0]]
+            return tf.cast(mask_val, labels.dtype) * tf.ones(
+                mask_shape, dtype=labels.dtype
+            )
+
+        def _assign_when_rows_not_empty():
+            targets = tf.gather(labels, match_indices)
+            if mask is None:
+                return targets
+            else:
+                masked_targets = tf.cast(mask_val, labels.dtype) * tf.ones_like(
+                    mask, dtype=labels.dtype
+                )
+                return tf.where(mask, masked_targets, targets)
+
+        return tf.cond(
+            tf.greater(num_gt_boxes, 0),
+            _assign_when_rows_not_empty,
+            _assign_when_rows_empty,
+        )
+
+    def _gather_batched(labels, match_indices, mask, mask_val):
+        """Gather based on batched labels."""
+        batch_size = labels.shape[0]
+        if batch_size == 1:
+            if mask is not None:
+                result = _gather_unbatched(
+                    tf.squeeze(labels, axis=0),
+                    tf.squeeze(match_indices, axis=0),
+                    tf.squeeze(mask, axis=0),
+                    mask_val,
+                )
+            else:
+                result = _gather_unbatched(
+                    tf.squeeze(labels, axis=0),
+                    tf.squeeze(match_indices, axis=0),
+                    None,
+                    mask_val,
+                )
+            return tf.expand_dims(result, axis=0)
+        else:
+            indices_shape = tf.shape(match_indices)
```

Comment:
It seems that the batched implementation is much simpler than the unbatched implementation. Shouldn't we just tf.expand_dim(inputs, 0) and let it always go through the batched implementation?

Reply:
I think the batched version is slightly more complicated? Otherwise the unbatched version is simple, but deals with one thing: empty rows

Reply:
readability width, I think the batched version is more concise, and the unbatched version as a lot of redirection and wrapped functions. And I would prefer to have single code path to handle both batch and unbatched case.

Reply:
discussed offline, the unbatched version needs to deal with empty rows, where the batched version do not, so keeping the code path as is.